test/hotspot/gtest/threadHelper.inline.hpp

```diff
@@ -117,6 +117,67 @@ class JavaTestThread : public JavaThread {
     t->main_run();
     t->_post->signal();
   }
+
+  void join() {
+    _post->wait();
+  }
+};
+
+// Calls a single-argument function of type F with state of type S as its input
+// in a new thread when doit() is run.
+template<typename F, typename S>
+class BasicTestThread : public JavaTestThread {
+private:
+  F _fun;
+  S _state;
+
+public:
+  BasicTestThread(F fun, S state, Semaphore* sem)
+    : JavaTestThread(sem),
+      _fun(fun),
+      _state(state) {
+  }
+
+  virtual ~BasicTestThread(){};
+
+  void main_run() override {
+    _fun(this, _state);
+  }
+};
+
+// A TestThreadGroup tracks multiple threads running the same function.
```

Comment:
>Could you please document what the template parameters are and what is expected of them?

>Does N really need to be a template parameter? You lose flexibility, and it is less readable too.

What flexibility are we losing? We're not going to be dynamically reallocate the amount of threads in a `TestThreadGroup`, or is this the exact use case you're imagining in the future?

I don't think it's necessarily less readable. It's fairly standard in C++ STL to use template arguments for sizes (see `std::array`). I know we typically use raw pointers and length variables in Hotspot, so it's certainly unidiomatic from those eyes. To me, it's more readable, because it's a hint to me that "aha, this size won't change during execution".

Reply:
The experience of developers with templates and STL will vary widely - just something to bear in mind when modernising the codebase. :) I found it harder to "spot the threads" in the new style.

Reply:
> The experience of developers with templates and STL will vary widely - just something to bear in mind when modernising the codebase. :) I found it harder to "spot the threads" in the new style.

Over-templatizing stuff is as bad as not using templates where they make sense. One negative effect, not the only one: it makes builds take longer, which is a tax every developer has to pay many times over. Not long ago, JDK builds used to be done in ~2.5 minutes on my fast machine, and now I'm up to almost 4 minutes. This, too, is death by a thousand cuts.

Other effects include APIs that are harder to read since template parameters are often undocumented. IDEs have a harder time coping. And code may become harder to re-use since what should be runtime decisions get baked into the code in the form of template args.

Reply:
Hi @jdksjolen,

> > So we have F, which is kind of clear, S, which is a state?, and then also a state generator? Which apparently needs to define operator () and return a state?
> 
> Let me explain my reasoning, call me out if something seems odd please!
> 
> A `BasicTestThread` has a F, the code which runs, and S, the local state which is only available for F. A ´TestThreadGroup`let's you create many`BasicTestThread` for the same F, but with varying S, with the help of a state generator.

Okay, I see. Thanks for clarifying. All examples atm hand out the same state to the individual threads though. Do we have many examples where we have different states?

One simpler, more pragmatic way would be for the thread group to give each started thread a numeric id. Most of the cases I can think of would be happy with that, using the id inside the thread function to control behavior.

> 
> Subclasses (the "old" way) are good at having their own local state, whilst lambdas are good at sharing state (through their access to the lexical environment they're created in). The whole S and state generator business is kind of a crutch to help lambdas out to be more like subclasses. Maybe there's a bit of YAGNI here, and really it's fine if `TestThreadGroup` also gets an S like `BasicTestThread` does?

Well, I don't see why this would have to be templatized at all; normal inheritance and interfaces would work just as well. At least with the examples given, I don't see the performance advantage. That's not to say that I'm against it, the in-function lambda syntax is certainly nice.

> 
> > Could you please document what the template parameters are and what is expected of them?
> 
> Absolutely. I'll also lengthen their name, to make them more clear-
> 
> > Does N really need to be a template parameter? You lose flexibility, and it is less readable too.
> 
> What flexibility are we losing? We're not going to be dynamically reallocate the amount of threads in a `TestThreadGroup`, or is this the exact use case you're imagining in the future?

The ability to start with the number of threads dictated at runtime, depending on runtime conditions. The ability to reuse the same TTG in different subtests with different thread numbers without template bloat.

Actually, the same argument goes for F. This design still requires us to use the same thread function for all threads. For cases where you want to have multiple different thread functions, you would still end up with some sort of switch inside the thread function. If F were a runtime thing, you could do something like this:

```
TTG.add_threads(12, functorX); // add 12 threads that do X
TTG.add_threads(6, functorY); // add 12 threads that do Y
```

i.e., add multiple threads with multiple functions to the same thread group and control them all via TTG.

> 
> I don't think it's necessarily less readable. It's fairly standard in C++ STL to use template arguments for sizes (see `std::array`). I know we typically use raw pointers and length variables in Hotspot, so it's certainly unidiomatic from those eyes. To me, it's more readable, because it's a hint to me that "aha, this size won't change during execution".

As I wrote in the other comment, opinions vary on this, and we have many devs who need to maintain the code base.

Reply:
>One simpler, more pragmatic way would be for the thread group to give each started thread a numeric id. Most of the cases I can think of would be happy with that, using the id inside the thread function to control behavior.

That might be sufficient. Looking through the test cases, they're either too complex for me, or would work OK with this kind of id. I'm thinking for N threads, each threads gets a unique id in the range `[0,N)`, are we thinking the same thing?

>The ability to start with the number of threads dictated at runtime, depending on runtime conditions. The ability to reuse the same TTG in different subtests with different thread numbers without template bloat.

There's probably no significant gain to be had in re-using the same TTG in different subtests, but I see the value of minimizing compilation times by having several monomorphizations depending on N.

>Actually, the same argument goes for F. This design still requires us to use the same thread function for all threads. For cases where you want to have multiple different thread functions, you would still end up with some sort of switch inside the thread function. If F were a runtime thing, you could do something like this:

That's flexible, but not easier to understand. Suddenly, several `doit()` and several `join()`s are viable, it makes the semantics of TTG way more complicated. Simple tests are preferable, I'd keep `F` if it meant that we *can't* add two different functors to a TTG.

We *can* get rid of `F`, as so:

```
#include <functional>

template<typename S>
class BasicTestThread : public JavaTestThread {
  std::function<void(Thread*, S)> f;
};
```

Unfortunately `std::function` is a no go, but it does show an enticing way forward. It would be useful in many places of Hotspot, probably having a real build time improvement.


Here's my plan:

- Move N out from being a template parameter to a constructor argument
- Try to remove S and instead use a numeric id, see how the two solutions differ for these test cases. If it's nice, I'll also remove S (and therefore also state generator, for example).
- I'm keeping F, no changes there.

Reply:
> > One simpler, more pragmatic way would be for the thread group to give each started thread a numeric id. Most of the cases I can think of would be happy with that, using the id inside the thread function to control behavior.
> 
> That might be sufficient. Looking through the test cases, they're either too complex for me to take the time to understand for this discussion, or would work OK with this kind of id. I'm thinking for N threads, each threads gets a unique id in the range `[0,N)`, are we thinking the same thing?

Yes. At least the tests I wrote, I usually had like one thread that would do A, the rest B, or half/half would do A or B. That can be just controlled via the running id. Beside its nice for debugging.

> 
> > The ability to start with the number of threads dictated at runtime, depending on runtime conditions. The ability to reuse the same TTG in different subtests with different thread numbers without template bloat.
> 
> There's probably no significant gain to be had in re-using the same TTG in different subtests, but I see the value of minimizing compilation times by not having several monomorphizations depending on N.

Not sure what you meant with the first half of the sentence. But one example for controlling num threads at runtime, I think now that we add more and more functionality to gtest, we may need a way to control the test width. Reduce the number of concurrent threads. There are folks running these tests as part of their build on tiny make machines. Of course, canonical test results would still require a certain number of concurrent tests.

> 
> > Actually, the same argument goes for F. This design still requires us to use the same thread function for all threads. For cases where you want to have multiple different thread functions, you would still end up with some sort of switch inside the thread function. If F were a runtime thing, you could do something like this:
> 
> That's flexible, but not easier to understand. Suddenly, several `doit()` and several `join()`s are viable, it makes the semantics of TTG way more complicated. Simple tests are preferable, I'd keep `F` if it meant that we _can't_ add two different functors to a TTG.

Ok, I agree.

> 
> We _can_ get rid of `F`, as so:
> 
> ```
> #include <functional>
> 
> template<typename S>
> class BasicTestThread : public JavaTestThread {
>   std::function<void(Thread*, S)> f;
> };
> ```
> 
> Unfortunately `std::function` is a no go, but it does show an enticing way forward. It would be useful in many places of Hotspot, probably having a real build time improvement.
> 
> Here's my plan:
> 
> * Move N out from being a template parameter to a constructor argument
> * Try to remove S and instead use a numeric id, see how the two solutions differ for these test cases. If it's nice, I'll also remove S (and therefore also state generator, for example).
> * I'm keeping F, no changes there.

Sounds good. Thanks for considering this.

Cheers, Thomas